### PR TITLE
[DO NOT MERGE] Track origin/suborigin of profile page views preliminary

### PR DESCRIPTION
**Pull request by yipingliao**
@2is10 cc @joshm1 

This is what I've come up with for tracking a page view that includes information about where it came from (origin and suborigin). I saw that @joshm1 had previously added a query parameter but I couldn't figure out how to take that off once we land on the new page (Josh doesn't take the query parameter off). Here I am using a service to hold the origin/suborigin data until it's used by tracking. It's a global state, basically, but it's what services are for. I made this into a directive so it can be easily popped into other places.

Please let me know what you think (and if you hate it).

Thanks!

### CODE REVIEW

File: kifi-backend/modules/shoebox/angular/src/userProfile/userProfile.js [hunk 1]
```diff
@@ -39,6 +39,8 @@ angular.module('kifi')
       }
       $rootScope.$emit('libraryUrl', {});
 
+      console.dir(originTrackingService.get());
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/userProfile/userProfile.js:42]: Whoever wants to use the data just calls the getter to retrieve the data.

File: kifi-backend/modules/shoebox/angular/src/userProfile/userProfileLibrariesList.tpl.html [hunk 2]
```diff
@@ -18,7 +18,7 @@
       </div>
     </a>
     <div itemscope itemtype="http://data-vocabulary.org/Person">
-      <a class="kf-upl-card-owner" href="{{lib.ownerProfileUrl}}" ng-click="trackProfileClick()" itemprop="url">
+      <a class="kf-upl-card-owner" href="{{lib.ownerProfileUrl}}" kf-track-origin origin="profile" sub-origin="curator" itemprop="url">
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/userProfile/userProfileLibrariesList.tpl.html:21]: Using this directive in just one place right now.

**Comment by joshm1** [comment]
Nice. I like this and prefer it to use query params. :+1:

File: kifi-backend/modules/shoebox/angular/src/userProfile/userProfileLibrariesList.tpl.html (hunk 2 above)
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/userProfile/userProfileLibrariesList.tpl.html:21]: Nice. Might consider a more concise single-attribute syntax:

``` html
kf-origin="profile"
kf-origin="profile/curator"
```

File: kifi-backend/modules/shoebox/angular/src/common/originTracking.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    return {
+      set: function (newOrigin) {
+        pageOrigin = _.cloneDeep(newOrigin);
+      },
+
+      get: function () {
+        var temp = pageOrigin;
+        pageOrigin = null;
+        return temp;
+      }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/common/originTracking.js:18]: It’s surprising for a method named `get` to be destructive. Might consider `getAndClear()` or `use()` or `take()`.

**Comment by 2is10** [comment]
I like this for in-app navigations. We’ll still need to use query or fragment parameters for entry into the app (e.g. from the browser extension). Might be simpler to just have the one method—the least common denominator—but it’s up to you. I guess the extension could also use message passing, but I’m afraid that would create a race condition.

File: kifi-backend/modules/shoebox/angular/src/userProfile/userProfileLibrariesList.tpl.html (hunk 2 above)
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/userProfile/userProfileLibrariesList.tpl.html:21]: Good idea. Thanks!

File: kifi-backend/modules/shoebox/angular/src/common/originTracking.js (hunk 3 above)
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/common/originTracking.js:18]: Will do.

**Comment by yipingliao** [comment]
We can have a wrapper method that gets the origin from the query parameters if they are available; and if they are not, retrieves them from the in-app service. I want to try to avoid having urls that have query parameters for tracking as much as possible. :)

**Pull request closed without merging**